### PR TITLE
Improvements for Danfoss Ally / Hive TRV

### PR DESCRIPTION
**Pull request by SwoopX**
- Enhanced the ability to steer/configure Danfoss Ally/Hive TRV thermostats via REST API
   - Expose _Viewing Direction (0x4000)_ from Thermostat UI configuration cluster as `config.displayflipped` (bool) for those 2 devices to change display orientation.
   - Expose _Keypad Lockout (0x0001)_ from Thermostat UI configuration cluster as `config.locked` (bool) for those 2 devices to set child lock.
   - Expose _Mounting mode active (0x4012)_ from Thermostat cluster as `state.mountingmodeactive` (bool) for those 2 devices to see if mounting mode is active.
   - Expose _Mounting mode control (0x4013)_ from Thermostat cluster as `config.mountingmode` (bool) for those 2 devices to set mounting mode.
   - Expose _SW error code (0x4000)_ from Diagnostics cluster as `config.errorcode` (string) for those 2 devices to obtain current error codes. Assumption is that value `0A00` inplies no error.
   - Added binding and repporting configuration for the required clusters and attributes as mentioned above.
- Added stricter data type check on bool for `config.displayflipped`, `config.locked` and `config.mountingmode`.
- Added `handleDiagnosticsClusterIndication` to handle data indications from diagnostics cluster.
- Added `handleThermostatUiConfigurationClusterIndication` to handle data indications from thermostat ui configuration cluster.

### CODE REVIEW

File: rest_sensors.cpp [hunk 1]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                        return REQ_READY_SEND;
                     }
+                }
+                else if (rid.suffix == RConfigDisplayFlipped)
+                {
+                    if (map[pi.key()].type() == QVariant::Bool)
+                    {
+                        if (sensor->modelId() == QLatin1String("eTRV0100") || sensor->modelId() == QLatin1String("TRV001"))
+                        {
+                            quint32 data = map[pi.key()].toUInt(&ok);
```
Comment (manup) [inline rest_sensors.cpp:1404]: Shouldn't `data` be used somewhere?

File: rest_sensors.cpp [hunk 2]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/displayflipped"), QString("Could not set attribute")));
+                                rsp.httpStatus = HttpStatusBadRequest;
+                                return REQ_READY_SEND;
+                            }
+                        }
+                    }
+                    else
                     {
-                        updated = true;
+                        rsp.list.append(errorToMap(ERR_INVALID_VALUE, QString("/sensors/%1/config/displayflipped"), QString("Invalid attribute value")));
```
Comment (manup) [inline rest_sensors.cpp:1420]: `QString("/sensors/%1/config/displayflipped")`

Is the `.arg(...)` missing for the %1 argument?

File: rest_sensors.cpp [hunk 3]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                            quint32 data = map[pi.key()].toUInt(&ok);
 
-                    if (SendTuyaRequest(task, TaskThermostat , 0x0107, data))
+                            if (addTaskThermostatUiConfigurationReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x4000, deCONZ::Zcl8BitEnum, data, VENDOR_DANFOSS))
+                            {
+                                updated = true;
+                            }
+                            else
+                            {
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/displayflipped"), QString("Could not set attribute")));
```
Comment (manup) [inline rest_sensors.cpp:1412]: `QString("/sensors/%1/config/displayflipped")`

Is the `.arg(...)` missing for the %1 argument?

File: rest_sensors.cpp [hunk 4]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                        {
+                            quint32 data = map[pi.key()].toUInt(&ok);
+
+                            if (addTaskThermostatUiConfigurationReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x0001, deCONZ::Zcl8BitEnum, data))
+                            {
+                                updated = true;
+                            }
+                            else
+                            {
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/locked"), QString("Could not set attribute")));
```
Comment (manup) [inline rest_sensors.cpp:1385]: `QString("/sensors/%1/config/locked")`

Is the `.arg(...)` missing for %1 argument?

File: rest_sensors.cpp [hunk 5]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                            }
+
+                            if (SendTuyaRequest(task, TaskThermostat , 0x0107, data))
+                            {
+                                updated = true;
+                            }
+                        }
+                        else if (sensor->modelId() == QLatin1String("eTRV0100") || sensor->modelId() == QLatin1String("TRV001"))
+                        {
+                            quint32 data = map[pi.key()].toUInt(&ok);
```
Comment (manup) [inline rest_sensors.cpp:1377]: Shouldn't `data` be used somewhere?

File: rest_sensors.cpp [hunk 6]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/locked"), QString("Could not set attribute")));
+                                rsp.httpStatus = HttpStatusBadRequest;
+                                return REQ_READY_SEND;
+                            }
+                        }
+                    }
+                    else
                     {
-                        data = QByteArray("\x01", 1);
+                        rsp.list.append(errorToMap(ERR_INVALID_VALUE, QString("/sensors/%1/config/locked"), QString("Invalid attribute value")));
```
Comment (manup) [inline rest_sensors.cpp:1393]: `QString("/sensors/%1/config/locked")`

Is the `.arg(...)` missing for %1 argument?

File: rest_sensors.cpp [hunk 7]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                        return REQ_READY_SEND;
+                    }
+                }
+                else if (rid.suffix == RConfigMountingMode)
+                {
+                    if (map[pi.key()].type() == QVariant::Bool)
+                    {
+                        if (sensor->modelId() == QLatin1String("eTRV0100") || sensor->modelId() == QLatin1String("TRV001"))
+                        {
+                            quint32 data = map[pi.key()].toUInt(&ok);
```
Comment (manup) [inline rest_sensors.cpp:1431]: Shouldn't `data` be used somewhere?

File: rest_sensors.cpp [hunk 8]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+
+                            if (addTaskThermostatUiConfigurationReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x4013, deCONZ::Zcl8BitEnum, data, VENDOR_DANFOSS))
+                            {
+                                updated = true;
+                            }
+                            else
+                            {
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/mountingmode"), QString("Could not set attribute")));
```
Comment (manup) [inline rest_sensors.cpp:1439]: `QString("/sensors/%1/config/mountingmode")`

Is the .arg(...) missing for the %1 argument?

File: rest_sensors.cpp [hunk 9]
```diff
@@ -1354,19 +1354,99 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                                rsp.httpStatus = HttpStatusBadRequest;
+                                return REQ_READY_SEND;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        rsp.list.append(errorToMap(ERR_INVALID_VALUE, QString("/sensors/%1/config/mountingmode"), QString("Invalid attribute value")));
```
Comment (manup) [inline rest_sensors.cpp:1447]: `QString("/sensors/%1/config/mountingmode")`

Is the `.arg(...)` missing for the %1 argument?

**Review by manup — CHANGES_REQUESTED** [review]
Added some comments for missing arguments and unhandled parameters.

File: rest_sensors.cpp (hunk 1 above)
Comment (SwoopX) [reply rest_sensors.cpp:1396]: It's used in the following function call.

File: rest_sensors.cpp (hunk 2 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp (hunk 3 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp (hunk 5 above)
Comment (SwoopX) [reply rest_sensors.cpp:1367]: It's used in the following function call.

File: rest_sensors.cpp (hunk 4 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp (hunk 6 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp (hunk 7 above)
Comment (SwoopX) [reply rest_sensors.cpp:1425]: It's used in the following function call.

File: rest_sensors.cpp (hunk 8 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp (hunk 9 above)
Comment (SwoopX) [reply rest_sensors.cpp]: Amended.

File: rest_sensors.cpp [hunk 10]
```diff
@@ -1344,19 +1344,105 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
+                        {
+                            quint32 data = map[pi.key()].toUInt(&ok);
+
+                            if (addTaskThermostatUiConfigurationReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x0001, deCONZ::Zcl8BitEnum, data))
+                            {
+                                updated = true;
+                            }
+                            else
+                            {
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/%2").arg(id).arg(pi.key()).toHtmlEscaped(),
```
Comment (manup) [inline rest_sensors.cpp:1375]: Out of curiosity why the `.toHtmlEscaped()`?

From the documentation:
> Converts a plain text string to an HTML string with HTML metacharacters <, >, &, and " replaced by HTML entities.

The strings are in a JSON object anyway, the only reason for escaping I can think of would be quotes or the other meta characters within the string, which we don't have.

**Review by manup — COMMENTED** [review]
Just a small question about the `QString::toHtmlEscaped()` usage, otherwise it looks fine.

**Pull request merged**